core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -271,56 +247,82 @@ private boolean unsubscribe(String queryName,
-     *
-     * @param uow           the Unit of Work in which the query handler will be invoked
-     * @param queryHandler  the query handler to be invoked
-     * @param updateHandler the update handler to be invoked with result of query handler
-     * @param <Q>           the query type
-     * @param <I>           the initial result type
-     * @param <U>           the incremental update type
-     * @return handle to cancel updates on this query
-     *
-     * @throws Exception propagated from query handler
-     */
-    private <Q, I, U> Registration invokeSubscriptionQueryHandler(UnitOfWork<QueryMessage<Q, I>> uow,
-                                                                  SubscriptionQueryMessageHandler<? super QueryMessage<?, I>, I, U> queryHandler,
-                                                                  UpdateHandler<I, U> updateHandler)
-            throws Exception {
-        SimpleQueryUpdateEmitter<I, U> emitter = new SimpleQueryUpdateEmitter<>(updateHandler);
-        I initialResult = interceptAndInvoke(uow, m -> queryHandler.handle(m, emitter)).getPayload();
-        emitter.initialResult(initialResult);
         return () -> {
-            emitter.cancelRegistration();
+            updateHandlers.remove(query);
             return true;
         };
     }
 
+    @SuppressWarnings("unchecked")
+    @Override
+    public <U> void emit(Predicate<SubscriptionQueryMessage<?, ?, U>> filter,
+                         Function<SubscriptionQueryMessage<?, ?, U>, U> update) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(sqm -> filter.test((SubscriptionQueryMessage<?, ?, U>) sqm))
+                              .map(m -> (SubscriptionQueryMessage<?, ?, U>) m)
+                              .forEach(query -> {
+                                  UpdateHandler<?, U> updateHandler = (UpdateHandler<?, U>) updateHandlers.get(query);
+                                  try {
+                                      U calculatedUpdate = update.apply(query);
+                                      updateHandler.onUpdate(calculatedUpdate);
+                                  } catch (Exception e) {
+                                      updateHandlers.remove(query);
+                                      updateHandler.onCompletedExceptionally(e);
+                                  }
+                              }));
+    }
+
+    @Override
+    public void complete(Predicate<SubscriptionQueryMessage<?, ?, ?>> filter) {
+        registeringSubscriptionQueryReadSafe(filter, () ->
+                updateHandlers.keySet()
+                              .stream()
+                              .filter(filter)
+                              .forEach(query -> {
+                                  UpdateHandler<?, ?> updateHandler = updateHandlers.get(query);
```

Comment:
Doesn't the `remove()` function return you the `UpdateHandler` which you want to call `onCompleted()` on?